### ISSUE TITLE
Axes3D voxels with alpha rendered wrong -- surfaces seem to cancel

### ISSUE THREAD

**Issue by lukasheinrich** [matplotlib/matplotlib#9745]
### Bug report

**Bug summary**

The shading of the voxels when using alpha looks off, when they overlap, shades seem to cancel each other out.

**Code for reproduction**

```
from mpl_toolkits.mplot3d import Axes3D

test_data = np.array(
      [[[ 0.005,  0.005],
        [ 0.000,  0.035]],

       [[ 0.005,  0.000],
        [ 0.005,  0.010]]])
alphas = np.divide(test_data,np.max(test_data))/2.0
```

**Actual outcome**

<img width="534" alt="screenshot" src="https://user-images.githubusercontent.com/2318083/32677515-ff79ee48-c65e-11e7-93a9-58a3c6710944.png">

**Expected outcome**

GIven the data source, I would have expected a result more like this, which was produced with the code found at 

http://stackoverflow.com/questions/30715083/python-plotting-a-wireframe-3d-cuboid?noredirect=1&lq=1

<img width="569" alt="screenshot" src="https://user-images.githubusercontent.com/2318083/32677557-359a107a-c65f-11e7-8e3a-517467659670.png">

**Matplotlib version**

  * Operating system: Mac OS X
  * Matplotlib version: 1.5.3
  * Matplotlib backend (`print(matplotlib.get_backend())`): module://ipykernel.pylab.backend_inline
  * Python version: Python 2.7.10
  * Jupyter version (if applicable): 1.0.0

**Comment by lukasheinrich** [comment]
a simpler test dataset exhibiting the same behavior is

```
np.array(
      [[[ 0.,  1],
        [ 0.,  2]],

       [[ 0.,  0],
        [ 0.,  1]]])
```

**Comment by jklymak** [comment]
Crossref #9559?

**Comment by WeatherGod** [comment]
no, this is different. The implementation drops faces for adjoining voxels. Perhaps we can make this optional?

**Comment by lukasheinrich** [comment]
@WeatherGod yes that would be great. the internal faces are important if you want to do a 3d `heatmap` / density plot.

**Comment by lukasheinrich** [comment]
maybe `internal_faces` ?

**Comment by WeatherGod** [comment]
Would `unexposed_faces` be clearer, perhaps?

**Comment by lukasheinrich** [comment]
to me `internal_faces` is more understandable. Also googling the term brings up similar problems, so maybe it's not a bad term?

http://meshlabstuff.blogspot.ch/2009/04/how-to-remove-internal-faces-with.html

**Comment by WeatherGod** [comment]
ok, `internal_faces` it is.

**Comment by lukasheinrich** [comment]
just following up, what's the status on this -- anyways we can contribute?

**Comment by WeatherGod** [comment]
Ah, yeah, I got it to partly work, but could only get 5 of the 6 faces to
get made for some reason. Then I ran out of time to work on it. How about I
push up my branch and set up a WIP PR so you can try it out and see what I
did wrong?

**Comment by WeatherGod** [comment]
@lukasheinrich , the PR is #9927

**Comment by eric-wieser** [comment]
This was by design, because faces are colored based on the color of the voxel they border. If a face borders two voxels, what color should it be?

**Comment by eric-wieser** [comment]
This was a deliberate decision in #6404, but it wasn't a principled one. There was a choice between

1. Drawing both faces in the same position, and letting z-fighting occur, making their color either a random function of viewing angle, or a function of draw order
2. Computing a "mixed" color for the two faces, and then drawing one. This is a little more work, and requires you to come up with a good mixing algorithm. Obviously for your single-color case, the algorithm is trivial.
3. Not drawing them at all

(1) could be fixed by drawing one-sided patches, but I'm not sure if matplotlib supports that.

**Comment by eric-wieser** [comment]
Note that you can use [the approach here](http://valdez.seos.uvic.ca/~jklymak/matplotlibhtml/gallery/mplot3d/voxels_numpy_logo.html) to force the culled faces to render, as well as avoiding z-fighting issues

**Comment by lukasheinrich** [comment]
@eric-wieser yeah I saw that example, but it seems like a lot of bookkeeping/data-shuffling for a result that (to me) seems like a natural expectation. Would people still be interested in the work started in #9927 ?

**Comment by eric-wieser** [comment]
I don't think #9927 is a good solution, because it overdraws two polygons of different colors in exactly the same spot

**Comment by lukasheinrich** [comment]
I do like the gapped look from the example you posted. Maybe this would be a good fit (perhaps with a adjustable `gapsize` kwarg?

**Comment by eric-wieser** [comment]
That might be reasonable, with `gapsize=None` or maybe `gapsize=0` being the current behavior. I think it should be a fraction of the voxel edge lengths, to scale well to the non-uniform cases.

**Comment by kbasye00** [comment]
Hi,  Any progress on this?  I was surprised to end up with something like this (where the two middle voxels are the same color):

![voxel_bug](https://user-images.githubusercontent.com/55330338/87232182-fa45c100-c38a-11ea-93f2-f44a9e8c098b.png)

If not, I might be interested in trying to work out the color blending based on https://en.wikipedia.org/wiki/Alpha_compositing#Alpha_blending   Thanks!!
